File: HelianzApi/Auth/AuthService.cs
using System.Data;
using Dapper;
using HelianzApi.Data;
using HelianzApi.Models;
using Microsoft.Extensions.Options;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Security.Cryptography;
using System.Text;

namespace HelianzApi.Auth;

public class AuthService
{
    private readonly DatabaseConnectionFactory _db;
    private readonly string _jwtKey;
    private readonly int _expiryHours;
    private readonly ILogger<AuthService> _logger;

    public AuthService(DatabaseConnectionFactory db, IConfiguration config, ILogger<AuthService> logger)
    {
        _db = db;
        _logger = logger;
        _jwtKey = config["Jwt:Key"] ?? "HelianzDevKey-ChangeInProduction-Min32Chars!";
        _expiryHours = int.TryParse(config["Jwt:ExpiryHours"], out var h) ? h : 24;
    }

    public async Task<LoginResponse?> LoginAsync(LoginRequest request)
    {
        try
        {
            using var conn = _db.CreateConnection();
            _logger.LogInformation("Login attempt for user: {User}", request.Username);

            // Look up user by username (matches OpenDental userod table)
            var user = await conn.QueryFirstOrDefaultAsync<UserRow>(@"
                SELECT UserNum, UserName, Password, UserGroupNum, ClinicNum,
                       EmployeeNum, IsHidden
                FROM userod
                WHERE UserName = @UserName AND IsHidden = 0",
                new { request.Username });

            if (user == null)
            {
                _logger.LogWarning("User not found: {User}", request.Username);
                return null;
            }

            // Verify password hash (OpenDental format: HashType$Salt$Hash)
            if (!VerifyPassword(request.Password, user.Password))
            {
                _logger.LogWarning("Invalid password for user: {User}", request.Username);
                return null;
            }

            // Get user's clinic access
            var clinicNums = (await conn.QueryAsync<long>(@"
                SELECT ClinicNum FROM userclinic WHERE UserNum = @UserNum",
                new { user.UserNum })).ToList();

        if (clinicNums.Count == 0)
            clinicNums.Add(user.ClinicNum);

        // Get user's group memberships
        var userGroupNums = (await conn.QueryAsync<long>(@"
            SELECT UserGroupNum FROM usergroupattach WHERE UserNum = @UserNum",
            new { user.UserNum })).ToList();

        // Query permissions for all groups the user belongs to
        var permissions = new List<UserPermission>();
        if (userGroupNums.Count > 0)
        {
            permissions = (await conn.QueryAsync<UserPermission>(@"
                SELECT DISTINCT gp.PermType, gp.FKey, gp.NewerDate, gp.NewerDays
                FROM grouppermission gp
                WHERE gp.UserGroupNum IN @UserGroupNums
                ORDER BY gp.PermType, gp.FKey",
                new { UserGroupNums = userGroupNums })).ToList();
        }

        // Build JWT token with permission claims
        var token = GenerateToken(user.UserNum, user.UserName, clinicNums, userGroupNums, permissions);

        return new LoginResponse
        {
            Token = token,
            DisplayName = user.UserName,
            UserNum = user.UserNum,
            ClinicNum = user.ClinicNum,
            ClinicNums = clinicNums,
            UserGroupNums = userGroupNums,
            Permissions = permissions
        };
    }
    catch (Exception ex)
    {
        _logger.LogError(ex, "Login failed for user {User}", request.Username);
        throw;
    }
}

    /// <summary>Generate a token without password verification (debug only).</summary>
    public string GenerateDebugToken(long userNum, string username, List<long> clinicNums)
        => GenerateToken(userNum, username, clinicNums, new(), new());

    private string GenerateToken(long userNum, string username, List<long> clinicNums,
        List<long> userGroupNums, List<UserPermission> permissions)
    {
        var claims = new List<Claim>
        {
            new(ClaimTypes.NameIdentifier, userNum.ToString()),
            new(ClaimTypes.Name, username),
        };
        claims.AddRange(clinicNums.Select(c => new Claim("ClinicNum", c.ToString())));
        claims.AddRange(userGroupNums.Select(g => new Claim("UserGroupNum", g.ToString())));
        // Store key permission types as claims for quick server-side checks
        foreach (var gp in permissions.GroupBy(p => p.PermType))
        {
            var fkeys = string.Join(",", gp.Select(p => p.FKey));
            claims.Add(new Claim($"Perm_{gp.Key}", fkeys));
        }

        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_jwtKey));
        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);

        var token = new JwtSecurityToken(
            claims: claims,
            expires: DateTime.UtcNow.AddHours(_expiryHours),
            signingCredentials: creds);

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    /// <summary>
    /// Verifies password against OpenDental hash format: HashType$Salt$Hash
    /// HashTypes: None, MD5, MD5_ECW, SHA3_512
    /// SHA3_512: Unicode(salt+pass) → SHA3-512 → Base64
    /// MD5:      Unicode(pass) → MD5 → Base64 (24 chars, ends with ==)
    /// MD5_ECW:  ASCII(pass) → MD5 → hex lowercase (32 chars)
    /// </summary>
    private static bool VerifyPassword(string plaintext, string storedHash)
    {
        if (string.IsNullOrEmpty(storedHash))
            return string.IsNullOrEmpty(plaintext);

        var parts = storedHash.Split('$');

        // Parse HashType$Salt$Hash format
        string hashType;
        string salt;
        string hash;

        if (parts.Length == 3 && Enum.TryParse<HashType>(parts[0], out _))
        {
            hashType = parts[0];
            salt = parts[1];
            hash = parts[2];
        }
        else if (storedHash.Length == 24 && storedHash.EndsWith("==") && !storedHash.Contains('$'))
        {
            // Legacy MD5 base64 hash (24 chars, no $ separator)
            hashType = "MD5";
            salt = "";
            hash = storedHash;
        }
        else
        {
            // Unknown format — treat as plain comparison
            return string.Equals(plaintext, storedHash);
        }

        string computedHash = hashType switch
        {
            "SHA3_512" => HashSHA3_512(salt + plaintext),
            "SHA512" => HashSHA3_512(salt + plaintext),
            "MD5" => HashMD5(plaintext),
            "MD5_ECW" => HashMD5_ECW(plaintext),
            "None" => plaintext,
            _ => plaintext
        };

        return ConstantEquals(computedHash, hash);
    }

    private static string HashSHA3_512(string input)
    {
        if (string.IsNullOrEmpty(input)) return "";
        var bytes = Encoding.Unicode.GetBytes(input);      // UTF-16 LE
        try
        {
            var hashBytes = System.Security.Cryptography.SHA3_512.HashData(bytes);
            return Convert.ToBase64String(hashBytes);
        }
        catch (PlatformNotSupportedException)
        {
            // Windows Server 2022 lacks CNG SHA3 support — use managed implementation
            using var sha3 = SHA3.Net.Sha3.Sha3512();
            var hashBytes = sha3.ComputeHash(bytes);
            return Convert.ToBase64String(hashBytes);
        }
    }

    private static string HashMD5(string input)
    {
        if (string.IsNullOrEmpty(input)) return "";
        var bytes = Encoding.Unicode.GetBytes(input);      // UTF-16 LE
        var hashBytes = System.Security.Cryptography.MD5.HashData(bytes);
        return Convert.ToBase64String(hashBytes);
    }

    private static string HashMD5_ECW(string input)
    {
        if (string.IsNullOrEmpty(input)) return "";
        var bytes = Encoding.ASCII.GetBytes(input);
        var hashBytes = System.Security.Cryptography.MD5.HashData(bytes);
        return Convert.ToHexString(hashBytes).ToLowerInvariant();
    }

    private static bool ConstantEquals(string a, string b)
    {
        if (a.Length != b.Length) return false;
        int diff = 0;
        for (int i = 0; i < a.Length; i++)
            diff |= a[i] ^ b[i];
        return diff == 0;
    }

    private enum HashType { None, MD5, MD5_ECW, SHA3_512 }

    private class UserRow
    {
        public long UserNum { get; set; }
        public string UserName { get; set; } = "";
        public string Password { get; set; } = "";
        public long UserGroupNum { get; set; }
        public long ClinicNum { get; set; }
        public long EmployeeNum { get; set; }
        public bool IsHidden { get; set; }
    }
}

File: helianz-api-node/src/services/authService.ts
import { createHash } from 'crypto';
import jwt from 'jsonwebtoken';
import pool from '../config/database';
import { LoginRequest, LoginResponse, UserPermission, JwtPayload } from '../models';
import pino from 'pino';

const logger = pino({ name: 'auth-service' });
const JWT_KEY = process.env.JWT_KEY || 'HelianzDevKey-ChangeInProduction-Min32Chars!';
const JWT_EXPIRY_HOURS = parseInt(process.env.JWT_EXPIRY_HOURS || '24', 10);

interface UserRow {
  UserNum: number;
  UserName: string;
  Password: string;
  UserGroupNum: number;
  ClinicNum: number;
  EmployeeNum: number;
  IsHidden: number;
}

/**
 * Verify password against OpenDental hash formats:
 *   SHA3_512: Unicode(salt+pass) → SHA3-512 → Base64
 *   MD5:      Unicode(pass) → MD5 → Base64 (24 chars, ends with ==)
 *   MD5_ECW:  ASCII(pass) → MD5 → hex lowercase (32 chars)
 *
 * Stored format: HashType$Salt$Hash
 */
function verifyPassword(plaintext: string, storedHash: string): boolean {
  if (!storedHash) return !plaintext;

  // Parse HashType$Salt$Hash
  const parts = storedHash.split('$');
  let hashType: string;
  let salt: string;
  let hash: string;

  if (parts.length === 3 || (parts.length === 2 && ['None', 'MD5', 'MD5_ECW', 'SHA3_512', 'SHA512'].includes(parts[0]))) {
    hashType = parts[0];
    salt = parts.length === 3 ? parts[1] : '';
    hash = parts[parts.length - 1];
  } else if (storedHash.length === 24 && storedHash.endsWith('==') && !storedHash.includes('$')) {
    // Legacy MD5 base64 hash (24 chars, no $ separator)
    hashType = 'MD5';
    salt = '';
    hash = storedHash;
  } else {
    return plaintext === storedHash;
  }

  let computed: string;
  switch (hashType) {
    case 'SHA3_512':
    case 'SHA512':
      computed = hashSHA3_512(salt + plaintext);
      break;
    case 'MD5':
      computed = hashMD5(plaintext);
      break;
    case 'MD5_ECW':
      computed = hashMD5_ECW(plaintext);
      break;
    case 'None':
      computed = plaintext;
      break;
    default:
      computed = plaintext;
  }

  return constantEquals(computed, hash);
}

function hashSHA3_512(input: string): string {
  if (!input) return '';
  const buf = Buffer.from(input, 'utf16le'); // UTF-16 LE = Unicode in .NET
  return createHash('sha3-512').update(buf).digest('base64');
}

function hashMD5(input: string): string {
  if (!input) return '';
  const buf = Buffer.from(input, 'utf16le');
  return createHash('md5').update(buf).digest('base64');
}

function hashMD5_ECW(input: string): string {
  if (!input) return '';
  const buf = Buffer.from(input, 'ascii');
  return createHash('md5').update(buf).digest('hex').toLowerCase();
}

function constantEquals(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

function generateToken(
  userNum: number,
  username: string,
  clinicNums: number[],
  userGroupNums: number[],
  permissions: UserPermission[]
): string {
  const claims: Record<string, string | string[]> = {
    sub: userNum.toString(),
    name: username,
    ClinicNum: clinicNums.map(String),
    UserGroupNum: userGroupNums.map(String),
  };

  // Store permission types as claims: Perm_<type> = "fkey1,fkey2,..."
  const permGroups = new Map<number, number[]>();
  for (const p of permissions) {
    const arr = permGroups.get(p.permType) || [];
    arr.push(p.fKey);
    permGroups.set(p.permType, arr);
  }
  for (const [permType, fkeys] of permGroups) {
    claims[`Perm_${permType}`] = fkeys.join(',');
  }

  const payload: Omit<JwtPayload, 'iat' | 'exp'> = {
    sub: userNum.toString(),
    name: username,
    ClinicNum: clinicNums.map(String),
    UserGroupNum: userGroupNums.map(String),
  };
  // Add dynamic Perm_ keys
  for (const [k, v] of Object.entries(claims)) {
    if (k.startsWith('Perm_')) {
      (payload as any)[k] = v;
    }
  }

  return jwt.sign(payload, JWT_KEY, {
    expiresIn: `${JWT_EXPIRY_HOURS}h`,
    algorithm: 'HS256',
  });
}

export async function loginAsync(request: LoginRequest): Promise<LoginResponse | null> {
  const conn = await pool.getConnection();
  try {
    logger.info({ user: request.Username }, 'Login attempt');

    const [users] = await conn.query<any[]>(
      `SELECT UserNum, UserName, Password, UserGroupNum, ClinicNum, EmployeeNum, IsHidden
       FROM userod WHERE UserName = ? AND IsHidden = 0`,
      [request.Username]
    );
    const user: UserRow | undefined = users[0];

    if (!user) {
      logger.warn({ user: request.Username }, 'User not found');
      return null;
    }

    if (!verifyPassword(request.Password, user.Password)) {
      logger.warn({ user: request.Username }, 'Invalid password');
      return null;
    }

    // Get clinic access
    const [clinicRows] = await conn.query<any[]>(
      `SELECT ClinicNum FROM userclinic WHERE UserNum = ?`,
      [user.UserNum]
    );
    const clinicNums: number[] = clinicRows.map((r: any) => r.ClinicNum);
    if (clinicNums.length === 0) clinicNums.push(user.ClinicNum);

    // Get user groups
    const [groupRows] = await conn.query<any[]>(
      `SELECT UserGroupNum FROM usergroupattach WHERE UserNum = ?`,
      [user.UserNum]
    );
    const userGroupNums: number[] = groupRows.map((r: any) => r.UserGroupNum);

    // Get permissions
    let permissions: UserPermission[] = [];
    if (userGroupNums.length > 0) {
      const placeholders = userGroupNums.map(() => '?').join(',');
      const [permRows] = await conn.query<any[]>(
        `SELECT DISTINCT gp.PermType, gp.FKey, gp.NewerDate, gp.NewerDays
         FROM grouppermission gp
         WHERE gp.UserGroupNum IN (${placeholders})
         ORDER BY gp.PermType, gp.FKey`,
        userGroupNums
      );
      permissions = permRows.map((r: any) => ({
        permType: r.PermType,
        name: '',
        fKey: r.FKey,
        newerDate: r.NewerDate,
        newerDays: r.NewerDays,
      }));
    }

    const token = generateToken(user.UserNum, user.UserName, clinicNums, userGroupNums, permissions);

    return {
      token,
      displayName: user.UserName,
      userNum: user.UserNum,
      clinicNum: user.ClinicNum,
      clinicNums,
      userGroupNums,
      permissions,
    };
  } catch (err) {
    logger.error({ err, user: request.Username }, 'Login failed');
    throw err;
  } finally {
    conn.release();
  }
}

export function generateDebugToken(userNum: number, username: string, clinicNums: number[]): string {
  return generateToken(userNum, username, clinicNums, [], []);
}
